In [1]:
import os
import warnings
import numpy as np
import pandas as pd
from sklearn import metrics
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV, StratifiedKFold, train_test_split
from sklearn.pipeline import FeatureUnion
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.svm import SVC

# titanic data

Najpierw zaloguj się do https://www.kaggle.com/ i przejdź do wyzwania https://www.kaggle.com/c/titanic, aby pobrać 
 * train.csv i test.csv. 

Zapisz je w katalogu data/titanic.

In [2]:
TITANIC_PATH = os.path.join("..", "data", "titanic")

In [3]:
def load_titanic_data(filename, titanic_path=TITANIC_PATH):
    csv_path = os.path.join(titanic_path, filename)
    return pd.read_csv(csv_path)

In [4]:
train_data = load_titanic_data("train.csv")
test_data = load_titanic_data("test.csv")

* Dane są już podzielone na zestaw treningowy i zestaw testów. 
* Jednak dane testowe nie zawierają etykiet: Twoim celem jest wyszkolenie najlepszego modelu, który możesz wykorzystać w danych treningowych, następnie dokonanie swoich przewidywań na danych testowych i przesłanie ich do Kaggle, aby zobaczyć ostateczny wynik.

Rzućmy okiem na kilka pierwszych rzędów zestawu treningowego:

In [5]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

The attributes have the following meaning:

* Survived: that's the target, 0 means the passenger did not survive, while 1 means he/she survived.
* Pclass: passenger class.
* Name, Sex, Age: self-explanatory
* SibSp: how many siblings & spouses of the passenger aboard the Titanic.
* Parch: how many children & parents of the passenger aboard the Titanic.
* Ticket: ticket id
* Fare: price paid (in pounds)
* Cabin: passenger's cabin number
* Embarked: where the passenger embarked the Titanic
* Let's get more info to see how much data is missing:

In [6]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Atrybuty **Age**, **Cabin** oraz **Embarked** są czasami zerowe (mniej niż 891 wartości bez wartości null), szczególnie w przypadku **Cabin** (77% ma wartość zerową). Zignorujemy teraz **Cabin** i skupimy się na reszcie. Atrybut **Age** ma około 19% wartości pustych, więc będziemy musieli zdecydować, co z nimi zrobić. Zastąpienie wartości null medianą wieku wydaje się uzasadnione.

Atrybuty **Name** i **Ticket** mogą mieć pewną wartość, ale będą one nieco trudne do przekształcenia w użyteczne liczby. Na razie będziemy je ignorować.

Rzućmy okiem na atrybuty liczbowe:

In [7]:
train_data = train_data.drop(["Name", "Ticket", "Cabin"], axis=1)  
test_data = test_data.drop(["Name", "Ticket", "Cabin"], axis=1)

In [8]:
train_data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Sex              0
Age            177
SibSp            0
Parch            0
Fare             0
Embarked         2
dtype: int64

In [9]:
train_data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

* Tylko 38% przeżyło: to wystarczająco blisko do 40%, więc **accuracy** będzie rozsądną miarą do oceny naszego modelu.

Sprawdźmy, czy etykiety przyjmują wartości 0 lub 1:

In [10]:
(train_data["Survived"].isin([0, 1])).all()

True

Nie zapomnij o etykietach:

In [11]:
y = train_data["Survived"]
X = train_data.drop("Survived", axis=1)

Teraz rzućmy okiem na wszystkie atrybuty kategoryczne:

In [12]:
X["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [13]:
X["Embarked"].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [14]:
# X["Cabin"].value_counts()

Atrybut **Embarked** mówi nam, gdzie pasażer wsiadł: C=Cherbourg, Q=Queenstown, S=Southampton.

Podzielmy zbiór na test/train:

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

Teraz zbudujmy nasze **pipeline** preprocessingu. 

Wykorzystamy DataframeSelector aby wybrać określone atrybuty z DataFrame:

In [16]:
# A class to select numerical or categorical columns 
# since Scikit-Learn doesn't handle DataFrames yet
class DataFrameSelector(BaseEstimator, TransformerMixin):
    def __init__(self, attribute_names):
        self.attribute_names = attribute_names

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        return X[self.attribute_names]

Zbudujmy **pipeline** dla atrybutów numerycznych:

In [17]:
imputer = SimpleImputer(strategy="median")

num_pipeline = Pipeline([
    ("select_numeric", DataFrameSelector(["Age"])),
    ("imputer", SimpleImputer(strategy="median")),
])

In [18]:
num_pipeline.fit_transform(X_train)

array([[45.5 ],
       [23.  ],
       [32.  ],
       [26.  ],
       [ 6.  ],
       [24.  ],
       [45.  ],
       [29.  ],
       [28.  ],
       [28.  ],
       [42.  ],
       [36.  ],
       [33.  ],
       [17.  ],
       [29.  ],
       [50.  ],
       [35.  ],
       [38.  ],
       [34.  ],
       [17.  ],
       [11.  ],
       [61.  ],
       [30.  ],
       [ 7.  ],
       [63.  ],
       [20.  ],
       [28.  ],
       [29.  ],
       [36.  ],
       [28.  ],
       [50.  ],
       [27.  ],
       [30.  ],
       [33.  ],
       [28.  ],
       [28.  ],
       [ 2.  ],
       [25.  ],
       [51.  ],
       [25.  ],
       [28.  ],
       [28.  ],
       [24.  ],
       [18.  ],
       [28.  ],
       [25.  ],
       [24.  ],
       [22.  ],
       [ 0.92],
       [24.  ],
       [26.  ],
       [34.  ],
       [21.  ],
       [28.  ],
       [28.  ],
       [28.  ],
       [22.  ],
       [62.  ],
       [18.  ],
       [42.  ],
       [57.  ],
       [19.  ],
       [

Będziemy także potrzebować imputera do kategorycznych kolumn  napisowych (zwykły Imputer nie działa na tych kolumnach):

In [19]:
# Inspired from stackoverflow.com/questions/25239958
class MostFrequentImputer(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        self.most_frequent_ = pd.Series([X[c].value_counts().index[0] for c in X],
                                        index=X.columns)
        return self

    def transform(self, X, y=None):
        return X.fillna(self.most_frequent_)

Teraz możemy zbudować **pipeline** dla atrybutów kategorycznych:

In [20]:
cat_pipeline = Pipeline([
    ("select_cat", DataFrameSelector(["Sex", "Embarked"])),
    ("imputer", MostFrequentImputer()),
    ("cat_encoder", OneHotEncoder(handle_unknown='ignore')),
])

In [21]:
cat_pipeline.fit_transform(X_train)

<712x5 sparse matrix of type '<class 'numpy.float64'>'
	with 1424 stored elements in Compressed Sparse Row format>

Na koniec połączmy powyższe podejścia:

In [22]:
preprocess_pipeline = FeatureUnion(transformer_list=[
    ("num_pipeline", num_pipeline),
    ("cat_pipeline", cat_pipeline),
])

Teraz mamy fajny **pipeline** przetwarzania wstępnego, który pobiera dane wejściowe i zwraca dane wyjściowe złorzone z liczb, które możemy podać do dowolnego modelu uczenia maszynowego.

# Zad

Robimy StratifiedKFold i znajdujemy optymalne parametry dla

* SVM z jądrem rbf
* SVM z jądrem poly
* SVM liniowego
* Regresji logistycznej

In [23]:
seed = 123
kfold = StratifiedKFold(n_splits=5, random_state=seed, shuffle=True)

In [24]:
def grid_fit(pipe, param_grid):
    grid = GridSearchCV(pipe, param_grid, cv=kfold, return_train_score=True)
    grid.fit(X_train, y_train)

    results = pd.DataFrame(grid.cv_results_)
    scores = np.array(results.mean_test_score)

    return scores, grid.best_estimator_, grid.best_params_

In [25]:
def score_model(model):
    y_pred = model.predict(X_test)

    precision = metrics.precision_score(y_test, y_pred)
    recall = metrics.recall_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred)
    accuracy = metrics.accuracy_score(y_test, y_pred)

    return precision, recall, f1, accuracy

In [26]:
param_grid = {
    'classifier__C': [0.001, 0.01, 0.1, 1, 10, 100]
}

methods = [
    ("SVM Linear", Pipeline([('preprocessing', preprocess_pipeline), ('classifier', SVC(kernel="linear"))]),
     param_grid),
    ("SVM RBF", Pipeline([('preprocessing', preprocess_pipeline), ('classifier', SVC(kernel="rbf"))])
     , param_grid),
    ("SVM Poly", Pipeline([('preprocessing', preprocess_pipeline), ('classifier', SVC(kernel="poly"))])
     , param_grid),
    ("Logistic Regression", Pipeline([('preprocessing', preprocess_pipeline), ('classifier', LogisticRegression())])
     , param_grid)
]


In [27]:
df = pd.DataFrame(columns=["name", "C", "precision", "recall", "f1", "accuracy"])

best_f1 = None
best_model = None

for i, (method, pipe, param_grid) in enumerate(methods):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        scores, best_model, best_params = grid_fit(pipe, param_grid)
        precision, recall, f1, accuracy = score_model(best_model)
        if best_f1 is None or f1 > best_f1:
            best_f1 = f1
            best_model = best_model
        df.loc[i] = [method, best_params['classifier__C'], *score_model(best_model)]

df

name       C  precision    recall        f1  accuracy
0           SVM Linear    0.01   0.753623  0.702703  0.727273  0.782123
1              SVM RBF   10.00   0.746479  0.716216  0.731034  0.782123
2             SVM Poly  100.00   0.714286  0.472973  0.569106  0.703911
3  Logistic Regression    0.10   0.753623  0.702703  0.727273  0.782123

Zajmijmy się danymi testowymi:

In [34]:
predictions = best_model.predict(test_data)
submission_df = pd.DataFrame({
    "PassengerId": test_data["PassengerId"],
    "Survived": predictions
})
submission_df.to_csv('submission.csv', index=False)

![Wynik](scr.png)